### ISSUE TITLE
changeTokenBalances runs transactions twice

### ISSUE THREAD

**Issue by Janther** [NomicFoundation/hardhat#2928]
I recently moved from `hardhat-waffle` to `hardhat-chai-matchers` and the matcher [changeTokenBalances](https://hardhat.org/hardhat-chai-matchers/docs/reference#.changetokenbalances) runs the transaction twice now.

```TypeScript
expect(await token.balanceOf(receiver)).to.equal(0); // initial value

await expect(token.transfer(receiver, 1000)).to.changeTokenBalances(
  token,
  [sender, receiver],
  [-1000, 1000]
); // Passes

expect(await token.balanceOf(receiver)).to.equal(1000); // Fails, somehow receiver now has 2000
```

**Comment by fvictorio** [comment]
Hey @Janther, I can't reproduce this with this code:

Mock token:

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.9;

contract Token {
  mapping(address => uint) public balanceOf;

  constructor() {
    balanceOf[msg.sender] = 1000000;
  }

  function transfer(address to, uint amount) public {
    balanceOf[msg.sender] -= amount;
    balanceOf[to] += amount;
  }
}
```

Test:

```js
const { expect } = require("chai");

it("test", async function () {
  const [sender, receiver] = await ethers.getSigners();
  const Token = await ethers.getContractFactory("Token");
  const token = await Token.deploy();

  await expect(token.transfer(receiver.address, 1000)).to.changeTokenBalances(
    token,
    [sender.address, receiver.address],
    [-1000, 1000]
  );

  expect(await token.balanceOf(receiver.address)).to.equal(1000);
});
```

**Comment by Janther** [comment]
aaaah I found the issue.

I was giving expect a function.

```Typescript
await expect(() => token.transfer(receiver.address, 1000)).to.changeTokenBalances(
  token,
  [sender.address, receiver.address],
  [-1000, 1000],
);
```

This is how it's shown in [ethereum-waffle documentation](https://ethereum-waffle.readthedocs.io/en/latest/matchers.html?highlight=changeBalance#change-token-balance-multiple-accounts).

when I migrated I missed the removal of the `() =>`

**Comment by Janther** [comment]
Maybe this can be added to the Documentation as it's not a drop-in replacement for this particular case.

**Comment by fvictorio** [comment]
Oh, I see, thanks for the clarification. We do want to support that use case for migration purposes, even if we don't want to encourage it (for consistency reasons; other matchers don't accept functions), so this should be considered a bug.

**Comment by fvictorio** [comment]
Fixed in [v1.0.2](https://github.com/NomicFoundation/hardhat/releases/tag/%40nomicfoundation%2Fhardhat-chai-matchers%401.0.2)
